From 4a683c529fecc354e52fbbb93ae17ea67fefcbf8 Mon Sep 17 00:00:00 2001
From: Jean-Christophe Trotin <jean-christophe.trotin@st.com>
Date: Tue, 3 Jan 2017 15:26:31 +0100
Subject: [PATCH 23/24] STM: waylandsink: use buffer video meta

The video info (stride and offset) are retrieved from the buffer
video meta if available, otherwise from the pool video alignment
option. So, even if there's no pool associated with the buffer, it's
possible to get the video info (stride and offset).

Warning: do not take this commit on gstreamer-1.10 as this bug is
fixed in the 1.10 version through the "waylandsink: Port to vmeta and
GstVideoFrame" commit. On gstreamer-1.10 we also need to revert our
ST custom patch "STM: waylandsink: set video alignment".

With these 2 patches (apply + revert), the stride is retrieved only
from the video meta, not from the pool.

https://stintbugzilla.st.com/show_bug.cgi?id=107485

Change-Id: I4a9c1b60500818404a22e20eb8be16526ac94218
Reviewed-on: https://gerrit.st.com/62466
Reviewed-by: Jean Christophe TROTIN <jean-christophe.trotin@st.com>
Tested-by: Jean Christophe TROTIN <jean-christophe.trotin@st.com>
---
 ext/wayland/gstwaylandsink.c | 16 ++++++++++++++--
 1 file changed, 14 insertions(+), 2 deletions(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index d21266a..91a14fb 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -617,6 +617,7 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
     /* No pool is proposed, but a minimum number of buffers and a size are
      * requested */
     gst_query_add_allocation_pool (query, NULL, info.size, 3, 0);
+    gst_query_add_allocation_meta (query, GST_VIDEO_META_API_TYPE, NULL);
   }
 
   return TRUE;
@@ -744,14 +745,25 @@ gst_wayland_sink_render (GstBaseSink * bsink, GstBuffer * buffer)
         "writing directly", buffer);
     to_render = buffer;
   } else {
+    GstVideoMeta *vmeta;
     GstMemory *mem;
     struct wl_buffer *wbuf = NULL;
 
     GST_LOG_OBJECT (sink, "buffer %p does not have a wl_buffer from our "
         "display, creating it", buffer);
 
-    /* Set video alignement */
-    gst_wayland_set_video_alignment (buffer->pool, &sink->video_info);
+    /* update video info from video meta */
+    vmeta = gst_buffer_get_video_meta (buffer);
+    if (vmeta) {
+      gint i;
+      for (i = 0; i < vmeta->n_planes; i++) {
+        sink->video_info.offset[i] = vmeta->offset[i];
+        sink->video_info.stride[i] = vmeta->stride[i];
+      }
+    } else if (buffer->pool) {
+      /* update video info from pool video alignment */
+      gst_wayland_set_video_alignment (buffer->pool, &sink->video_info);
+    }
 
     mem = gst_buffer_peek_memory (buffer, 0);
 
-- 
2.7.4

